### PR TITLE
COM-394: Automatically generate `className` for component slots with `createComponentSlot()`

### PR DESCRIPTION
**Pull request by jamesricky**
E.g. the `Root` slot of `MainContent` will now alway the className `CometAdminMainContent-root`. 
If the `fullHeight` prop is set to true, the className `CometAdminMainContent-fullHeight` also will be added. 

This was previously done automatically, when using `withStyles()` from the legacy `@mui/styles` package, which was removed during the theming refactor: #1376

This now matches the behavior of Comet-Admin components before the theming refactor and MUI components.

### CODE REVIEW

**Comment by jamesricky** [comment]
@johnnyomair FYI I'm not really sure how to fix the types in the lines with the `@ts-expect-error` in `createSlot.tsx`...

However the usage of the `createSlot` function infers the types of the base component correctly, e.g. all props of MUIs `Paper` are available with `const MyPaper = createSlot(MuiPaper)`, all props specific to the `a` tag are available with `createSlot("a")` and so on.

File: packages/admin/cms-admin/src/pages/pagesPage/PagesPage.tsx [hunk 1]
```diff
@@ -158,7 +158,7 @@ export function PagesPage({
                         </Toolbar>
                         <PageTreeContext.Provider value={{ allCategories, currentCategory: category, documentTypes, tree, query: pagesQuery }}>
                             <MainContent fullHeight sx={{ display: "flex", flexDirection: "column" }}>
-                                <Box sx={{ width: "100%", paddingLeft: 24, paddingRight: 50 }}>
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/pages/pagesPage/PagesPage.tsx:161]: Is it safe to remove those?

File: packages/admin/blocks-admin/src/blocks/common/AdminComponentButton.tsx [hunk 2]
```diff
@@ -16,14 +16,18 @@ interface Props {
...
-const Button = styled(MuiButton)<Pick<Props, "size">>`
-    ${({ size, theme }) =>
-        size === "large" &&
+type ButtonProps = {
+    spacingSize?: Props["size"];
+};
+
+const Button = styled(MuiButton)<ButtonProps>`
+    ${({ spacingSize, theme }) =>
+        spacingSize === "large" &&
```
[the remark this answers is not in the document]
Comment (jamesricky) [reply packages/admin/blocks-admin/src/blocks/common/AdminComponentButton.tsx:30]: `size` is an existing prop from MUIs `Button` that this button extends. 
We don't want to set the value of that prop, only the value of the prop, that changes the padding of the component. 

To make this more understandable, I'll rename the prop to something else and change it to a boolean.

File: docs/docs/comet-core-development/create-admin-components-with-theme-support.md [hunk 3]
```diff
@@ -64,10 +64,10 @@ E.g.: in this instance of `MyComponent`, the root should have a red border, and
 See this as a working example in our [Storybook](https://storybook.comet-dxp.com/?path=/story/comet-admin-theming--themable-mycomponent).
 
 ```tsx
-import { ThemedComponentBaseProps } from "@comet/admin";
+import { ThemedComponentBaseProps, createSlot } from "@comet/admin";
```
[the remark this answers is not in the document]
Comment (jamesricky) [reply docs/docs/comet-core-development/create-admin-components-with-theme-support.md:67]: Good point, will change that.

**Comment by johnnyomair** [comment]
> @johnnyomair FYI I'm not really sure how to fix the types in the lines with the `@ts-expect-error` in `createSlot.tsx`...

We should pair on this. I can't determine the errors by reading the code alone. 😁

File: packages/admin/admin/src/common/toolbar/backbutton/ToolbarBackButton.tsx [hunk 4]
```diff
@@ -1,40 +1,39 @@
...
+})(
+    ({ theme }) => css`
+        flex: 0;
+        display: flex;
+        align-items: stretch;
+
+        .CometAdminToolbarItem-root {
+            padding: 0;
+            padding-right: ${theme.spacing(3)};
+        }
```
[the remark this answers is not in the document]
Comment (jamesricky) [reply packages/admin/admin/src/common/toolbar/backbutton/ToolbarBackButton.tsx:25]: They were accidentally removed in the theming-refactor, they already exist in `next`: https://github.com/vivid-planet/comet/blob/next/packages/admin/admin/src/common/toolbar/backbutton/ToolbarBackButton.tsx#L21

File: packages/admin/admin/src/table/TableDndOrder.tsx [hunk 5]
```diff
@@ -112,39 +113,30 @@ function DndOrderRow<TRow extends IRow>(props: IDndOrderRowProps<TRow>) {
...
 
-const DragItemContainer = styled("div", {
-    name: "CometAdminTableDndOrder",
-    slot: "dragItemContainer",
-    overridesResolver(_, styles) {
-        return [styles.dragItemContainer];
-    },
+const DragIconContainer = createComponentSlot("div")<TableDndOrderClassKey>({
+    componentName: "TableDndOrder",
+    slotName: "dragIconContainer",
```
[the remark this answers is not in the document]
Comment (jamesricky) [reply packages/admin/admin/src/table/TableDndOrder.tsx:139]: No, this was accidentally renamed from `Icon` to `Item` during the theming refactor, it's now the same again, as in `next`: https://github.com/vivid-planet/comet/blob/next/packages/admin/admin/src/table/TableDndOrder.tsx#L161

File: packages/admin/blocks-admin/src/blocks/common/AdminComponentButton.tsx [hunk 6]
```diff
@@ -16,14 +16,18 @@ interface Props {
...
-const Button = styled(MuiButton)<Pick<Props, "size">>`
-    ${({ size, theme }) =>
-        size === "large" &&
+type ButtonProps = {
+    increasedPadding: boolean;
+};
+
+const Button = styled(MuiButton)<ButtonProps>`
+    ${({ increasedPadding, theme }) =>
+        increasedPadding &&
```
[the remark this answers is not in the document]
Comment (jamesricky) [reply packages/admin/blocks-admin/src/blocks/common/AdminComponentButton.tsx:30]: Good point 👍🏼

File: packages/admin/admin-color-picker/src/ColorPicker.slots.ts [hunk 7]
```diff
@@ -255,20 +227,18 @@ export type PreviewIndicatorColorProps = {
...
+    slotName: "previewIndicator",
+    classesResolver({ type }) {
         return [
-            styles.previewIndicator,
-            type === "color" && styles.previewIndicatorColor,
-            type === "empty" && styles.previewIndicatorEmpty,
-            type === "invalid" && styles.previewIndicatorInvalid,
+            type === "color" && "previewIndicatorColor",
+            type === "empty" && "previewIndicatorEmpty",
+            type === "invalid" && "previewIndicatorInvalid",
```
[the remark this answers is not in the document]
Comment (jamesricky) [reply packages/admin/admin-color-picker/src/ColorPicker.slots.ts:237]: No, `classesResolver()` must return an array of the given `ClassKey` type. 
And `type` is typed from the `PreviewIndicatorProps` type in this case.

File: packages/admin/admin/src/helpers/createComponentSlot.tsx [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    classesResolver?: (ownerState: OwnerState) => (ClassKey | false | undefined)[];
+};
+
+type OwnerStateObjectIfDefined<OwnerState extends object | undefined> = OwnerState extends object
+    ? {
+          ownerState: OwnerState;
+      }
+    : {
+          ownerState?: undefined;
+      };
```
[the remark this answers is not in the document]
Comment (jamesricky) [reply packages/admin/admin/src/helpers/createComponentSlot.tsx:19]: This it to make sure the `ownerState` prop is required on slots, that include an `OwnerState` type/generic in their definition, and to make sure the prop is not set, when no `OwnerState` is included.

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**